Pull request for branch "issue76" (#76)

mcjazzyfunky opened the pull request:
@thysultan 
As discussed: Support for "contextTypes" has not been implemented - you've said you take care of that on your own.
Could you please review my changes?
Please let me know in case I should change/add/fix something.
Thanks.

Zolmeister commented on src/Core/Component.js:
```diff
@@ -465,5 +470,44 @@ function enqueueStateCallback (element, owner, callback) {
+		for (var propName in propTypes) {
+			var error = propTypes[propName](props, propName, componentName, 'prop', null)
+
+			if (error)
```
could add a check for `Error` type for enforcing correctness

mcjazzyfunky replied:
@thysultan @Zolmeister 
Hey guys, what exactly do you want now? .. as you gave me different task descriptions regarding the error check ;-)

Option 1: if (error instanceof Error) // => as given in the first example snippet by @thysultan 
Option 2: if (error && typeof error.message === 'string') // => Quote: "Error-like to support cross realm instances"
Option 3: if (error) // => Quote:  "maybe this is enough(consider any truthy value as cause for exception)"

I personally prefer Option 1 as I cannot think of any use case where the "cross realm capability" regarding that error object is needed ... but maybe that's just due to lack of imagination ;-)

Zolmeister replied:
This is what I meant:

```js
if (error) {
    if (error instanceof Error) {
        throw error
    else {
        throw new Error 'Return value must be of type `Error`'
    }
}
```

Zolmeister commented on test/Context.spec.js:
```diff
@@ -819,5 +819,19 @@ describe('Context', () => {
+			propTypes1 = {},
+			propTypes2 = {}
+
+		TestCtx1.Provider.propTypes = propTypes1
```
consider using [Object.getOwnPropertyDescriptor()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyDescriptor) instead

mcjazzyfunky replied:
@Zolmeister Mmmh, not really sure whether we really should use ```getPropertyDescriptor``` in the unit tests for this: The fact that the "propTypes" attribute is handled by getter and setter functions internally is just an implementation detail and therefore the aspect how the "propTypes" property is actually handled internally is not part of the public API (may change in future). As the unit tests do only test for the public API we should not use ```getOwnPropertyDescriptor``` IMHO....

Zolmeister replied:
I see, this is testing `set` and `get` methods, not attribute writeability. disregard

Zolmeister commented on test/Element.spec.js:
```diff
@@ -106,5 +106,84 @@ describe('Element', () => {
+
+	it('should validate propTypes in development envoironment and ignore in production environment', () => {
+		const
+			FunctionalComponent1 = ({ name }) => name,
```
do we not lint bracket spacing? (`{name}` vs `{ name }`)

Zolmeister commented on test/Element.spec.js:
```diff
@@ -106,5 +106,84 @@ describe('Element', () => {
+		if (process.env.NODE_ENV === 'development') {
+			assert.throws(task1, 'Validation failed for prop "name" of SFComponent1 (location: prop, fullPropName: null, props: {"name":"Jane Doe"}')
+			assert.doesNotThrow(task2)
+			assert.doesNotThrow(task3)
```
we should also test that `ClassBasedComponent` throws

mcjazzyfunky commented:
@thysultan @Zolmeister Okay, I've fixed all issues discovered at @Zolmeister's review. The whole task is about making development with DIO more comfortable, so I pimped the error messages a bit to be also a bit "more comfortable".

thysultan commented:
@mcjazzyfunky Lets move checkPropTypes out of Component into it's own file. i.e Types/Check?

thysultan commented on src/Core/Context.js:
```diff
@@ -154,20 +154,53 @@ ContextConsumer[SharedSitePrototype] = ObjectCreate(Component[SharedSitePrototyp
  * @public
  */
 function createContext (value) {
+	var CustomContextProvider, Provider, Consumer
```
Lets leave theses updates to `createContext` for another PR that also adds `contextTypes`.

thysultan commented on src/Core/PropTypes.js:
```diff
@@ -0,0 +1,61 @@
+        if (error instanceof Error) {
+          throw error
+        } else {
+          throw new Error(
```
We can leave this out to allow `return check[...] && Error(...)` pattern, that might return false.

thysultan commented on src/Core/PropTypes.js:
```diff
@@ -0,0 +1,61 @@
+      if (typeof validator !== 'function') {
+        throw new Error(
+          'Validator for propTypes.'
+            + propName
```
Just `Invalid propType validator` would suffice as error boundaries will generate a complete stack trace when this error propagates.

mcjazzyfunky commented:
@thysultan 
I've commited all requested changes (and some more) to branch "issue76":

- Removed propTypes logic for ContextProvider (file "Context.js" is equal to "master" now)
- If a propType validator is not a function there will not be any special logic any longer => instead, a "normal" "is not a function" exception will be thrown
- Modified propTypes validator handling: Any result value is allowed now => true is okay, false will throw an error and everything else will also throw an error object if Boolean(result) === true
- Moved propTypes related unit tests to new file "PropTypes.spec.js"
- Fixed #81 and added two corresponding unit tests

thysultan commented on test/Context.spec.js:
```diff
@@ -819,5 +819,19 @@ describe('Context', () => {
 		assert.html(container, `<button style="color: orange;">2</button><button>Colors!</button>`)
+	}),
+
+	it('should write and read propTypes of context providers properly', () => {
```
We can add this to `PropTypes.spec.js` when we update context to support `propTypes`.

mcjazzyfunky replied:
Uuups, thanks for the info ... I've completely forgotten that unit test :-( ... is removed

thysultan commented:
LGTM, feel free to merge this. Thanks.
